compiler/rustc_data_structures/src/sorted_map/index_map.rs

```diff
@@ -94,13 +94,15 @@ impl<I: Idx, K: Ord, V> SortedIndexMultiMap<I, K, V> {
         Q: Ord + ?Sized,
         K: Borrow<Q>,
     {
-        // FIXME: This should be in the standard library as `equal_range`. See rust-lang/rfcs#2184.
         match self.binary_search_idx(key) {
             Err(_) => self.idxs_to_items_enumerated(&[]),
 
             Ok(idx) => {
-                let start = self.find_lower_bound(key, idx);
-                let end = self.find_upper_bound(key, idx);
+                let start = self.idx_sorted_by_item_key[..idx]
+                    .partition_point(|&i| self.items[i].0.borrow() != key);
+                let end = idx
+                    + self.idx_sorted_by_item_key[idx..]
+                        .partition_point(|&i| self.items[i].0.borrow() == key);
                 self.idxs_to_items_enumerated(&self.idx_sorted_by_item_key[start..end])
```

Comment:
If I understand correctly, this is now doing a `binary_search_by` to use a binary search to find *any* matching key, and then does two more binary searches (left and right of the item that was found), to find the first one. Why not just use a `partition_point(|| .. < ..)` to find the first key directly?

That is, if you just remove the match and the Err case, and change the `!=` to ` <`, it still works and does the same thing, right?

Reply:
(And unless there's tons of duplicated keys, it can be a lot faster to only use partition_point for finding the lower bound, and just search linearly for the upper bound. See also the comment on line 128 before the change.)

Reply:
Oh neat, indeed we could improve it more. Almost I agree but I think we couldn't remove `amtch` for the case when the key doesn't exist.
Anyway, I'll open a PR to follow up it, it'd be great if you could review it, thanks!